framework/include/utils/ValueCache.h

```diff
@@ -62,34 +79,72 @@ class ValueCache
                                                  std::size_t>;
 
   std::unique_ptr<KdTreeT> _kd_tree;
-  std::vector<T> _data;
 
   const std::size_t _in_dim;
   const std::size_t _max_leaf_size;
   const std::size_t _max_subtrees;
+
+  std::optional<std::string> _persistent_storage_file;
+
+  friend void dataStore<T>(std::ostream & stream, ValueCache<T> & c, void * context);
+  friend void dataLoad<T>(std::istream & stream, ValueCache<T> & c, void * context);
 };
 
 template <typename T>
 ValueCache<T>::ValueCache(std::size_t in_dim, std::size_t max_leaf_size)
-  : _point_cloud(in_dim),
+  : _cache(in_dim),
     _kd_tree(nullptr),
     _in_dim(in_dim),
     _max_leaf_size(max_leaf_size),
     _max_subtrees(100)
 {
 }
 
+template <typename T>
+ValueCache<T>::ValueCache(const std::string & file_name,
+                          std::size_t in_dim,
+                          std::size_t max_leaf_size)
+  : ValueCache(in_dim, max_leaf_size)
+{
+  _persistent_storage_file = file_name;
+
+  // if the persistent storage file exists and is readable, load it
+  if (MooseUtils::checkFileReadable(*_persistent_storage_file,
+                                    /*check_line_endings =*/false,
+                                    /*throw_on_unreadable =*/false))
+  {
+    std::ifstream in_file(_persistent_storage_file->c_str());
+    if (!in_file)
+      mooseError("Failed to open '", *_persistent_storage_file, "' for reading.");
+    dataLoad(in_file, *this, nullptr);
+
+    // build kd-tree
+    _kd_tree = std::make_unique<KdTreeT>(_in_dim, _cache, _max_leaf_size);
+    mooseAssert(_kd_tree != nullptr, "KDTree was not properly initialized.");
+  }
+}
+
+template <typename T>
+ValueCache<T>::~ValueCache()
```

Comment:
When do these get destructed in actual use? Storing to file in a destructor sounds like it _could_ be dangerous

Reply:
Nah, destructors are guaranteed to fully run upon application exit. If you need the cache to survive unplanned crashed or HPC wall time expiry you need to make it restartable, which should now work, too.